src/data/fun_like/equiv.lean

```diff
@@ -0,0 +1,183 @@
+  extends equiv_like F A (λ _, B), my_hom_class F A B.
+
+-- You can replace `my_iso.equiv_like` with the below instance, or keep both:
+instance : my_iso_class (my_iso A B) A B :=
+{ coe := my_iso.to_fun,
+  inv := my_iso.inv_fun,
+  left_inv := my_iso.left_inv,
+  right_inv := my_iso.right_inv,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  map_op := my_iso.map_op' }
+
+-- [Insert `has_coe_to_fun`, `to_fun_eq_coe`, `ext` and `copy` here]
+```
+
+The second step is to add instances of your new `my_iso_class` for all types extending `my_iso`.
+Typically, you can just declare a new class analogous to `my_iso_class`:
+
+```
+structure cooler_iso (A B : Type*) [cool_class A] [cool_class B]
+  extends my_iso A B :=
+(map_cool' : to_fun cool_class.cool = cool_class.cool)
+
+class cooler_iso_class (F : Type*) (A B : out_param $ Type*) [cool_class A] [cool_class B]
+  extends my_iso_class F A B :=
+(map_cool : ∀ (f : F), f cool_class.cool = cool_class.cool)
+
+@[simp] lemma map_cool {F A B : Type*} [cool_class A] [cool_class B] [cooler_iso_class F A B]
+  (f : F) : f cool_class.cool = cool_class.cool :=
+my_iso_class.map_op
+
+-- You can also replace `my_iso.equiv_like` with the below instance:
+instance : cool_iso_class (cool_iso A B) A B :=
+{ coe := cool_iso.to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  map_op := cool_iso.map_op',
+  map_cool := cool_iso.map_cool' }
+
+-- [Insert `has_coe_to_fun`, `to_fun_eq_coe`, `ext` and `copy` here]
+```
+
+Then any declaration taking a specific type of morphisms as parameter can instead take the
+class you just defined:
+```
+-- Compare with: lemma do_something (f : my_iso A B) : sorry := sorry
+lemma do_something {F : Type*} [my_iso_class F A B] (f : F) : sorry := sorry
+```
+
+This means anything set up for `my_iso`s will automatically work for `cool_iso_class`es,
+and defining `cool_iso_class` only takes a constant amount of effort,
+instead of linearly increasing the work per `my_iso`-related declaration.
+
+-/
+
+/-- The class `equiv_like E α β` expresses that terms of type `E` have an
+injective coercion to functions from `α` to `β`.
+
+This typeclass is used in the definition of the homomorphism typeclasses,
+such as `zero_hom_class`, `mul_hom_class`, `monoid_hom_class`, ....
+-/
+class equiv_like (E : Sort*) (α β : out_param Sort*) :=
+(coe : E → α → β)
+(inv : E → β → α)
+(left_inv  : ∀ e, function.left_inverse (inv e) (coe e))
+(right_inv : ∀ e, function.right_inverse (inv e) (coe e))
+-- The `inv` hypothesis makes this easier to prove with `congr'`
+(coe_injective' : ∀ e g, coe e = coe g → inv e = inv g → e = g) 
+
+namespace equiv_like
+
+variables {E F α β γ : Sort*} [iE : equiv_like E α β] [iF : equiv_like F β γ]
+include iE
+
+lemma coe_injective : function.injective (equiv_like.coe : E → (α → β)) :=
+λ e g h, coe_injective' e g h ((left_inv e).eq_right_inverse (h.symm ▸ right_inv g))
```

Comment:
I would be inclined to inline this into `to_embedding_like` below, since:

* This statement is about `equiv_like.coe e`, which isn't the preferred spelling of `⇑e`
* With the correct spelling, this is `fun_like.coe_injective`